src/keys.ts

```diff
@@ -18,18 +18,63 @@
  * @category Object
  */
 
+import { IterableContainer } from './_types';
+
 export function keys(
   source: Record<PropertyKey, unknown> | ArrayLike<unknown>
 ): Array<string> {
   return Object.keys(source);
 }
 
+type Strict = <T extends object>(source: T) => Keys<T>;
+type Keys<T> = T extends IterableContainer ? ArrayKeys<T> : Array<ObjectKey<T>>;
+
+// The keys output can mirror the input when it is an array/tuple. We do this by
+// "mapping" each item "key" (which is actually an index) as its own value. This
+// would maintain the shape, even including labels.
+type ArrayKeys<T extends IterableContainer> = {
+  -readonly [Index in keyof T]: Index extends string | number
+    ? // Notice that we coallesce the values as strings, this is because in JS,
+      // Object.keys always returns strings, even for arrays.
+      `${IsIndexAfterSpread<T, Index> extends true ? number : Index}`
+    : // Index is typed as a symbol, this can't happen, but we need to guard
+      // against it for typescript.
+      never;
+};
+
+type IsIndexAfterSpread<
+  T extends IterableContainer,
+  Index extends string | number
+> = IndicesAfterSpread<T> extends never
+  ? false
+  : Index extends `${IndicesAfterSpread<T>}`
+  ? true
+  : false;
+
+// Find the index of the tuple where a spread item is located, and return all
+// indices in the tuple which are located after it. The tuple could be prefixed
+// by any number of literal items. If the input is a simple array it would
+// return 0 (as expected), and if the tuple doesn't contain a spread element it
+// would return never.
+type IndicesAfterSpread<
+  T extends ReadonlyArray<unknown> | [],
+  // We use this type to count how many items we consumed, it's just a pseudo-
+  // element that is used for its length.
+  Iterations extends ReadonlyArray<unknown> = []
+> = T[number] extends never
+  ? never
+  : T extends readonly [unknown, ...infer Tail]
+  ? IndicesAfterSpread<Tail, [unknown, ...Iterations]>
+  : T extends readonly [...infer Head, unknown]
+  ? IndicesAfterSpread<Head, [unknown, ...Iterations]> | Iterations['length']
+  : Iterations['length'];
+
+type ObjectKey<T> = {
+  [K in keyof T]-?: K extends string | number ? `${K}` : never;
+}[keyof T];
+
 export namespace keys {
-  export function strict<T extends Record<PropertyKey, unknown>>(
-    source: T
-  ): Array<
-    { [K in keyof T]-?: K extends string | number ? `${K}` : never }[keyof T]
-  > {
-    return keys(source) as any;
-  }
+  // @ts-expect-error [ts2322] - I don't know why i'm getting this, the typing
```

Comment:
hmm, we've had the experience in the past that these comments get stripped out during building, and then bring problems to people that don't have `skipLibCheck` turned on.

That was reported here:
- https://github.com/remeda/remeda/issues/163

Reply:
I've created #306 so that we can test our output files and avoid these issues without catching them in CR. Let's ship that and rebase and check it out

Reply:
I've rebased on top of #306 and the build is passing fine, I don't know though if that check is enough now, and I'm not sure I know how to proceed.

The problem is that this ts-expect-error is on the assignment to the strict variable, which has no bearing on the output type definitions, and the output logic itself is in JS so doesn't care about ts issues. I don't know how to simulate the skipLibCheck scenario here to see if it's relevant or not.

also, as far as I understand skipLibCheck is `false` by default, including in our own build ci tests, so I //think// we are fine here? 🤷🏼

Reply:
I also see no error, using the built remeda version from this PR (codesandbox preview build) in a project that has `skipLibCheck: false`. I think this means we're good ...